src/Microsoft.AspNet.OData.Shared/EdmChangedObjectCollection.cs

```diff
@@ -60,6 +70,316 @@ private void Initialize(IEdmEntityType entityType)
+                    }
+                }
+            }
+        }
+
+        private IEdmChangedObject HandleFailedOperation(EdmEntityObject changedObj, DataModificationOperationKind operation, IEdmStructuredObject originalObj,
+            IEdmStructuralProperty[] keys, string errorMessage, EdmODataAPIHandler apiHandler)
+        {
+            IEdmChangedObject edmChangedObject = null;
+            DataModificationExceptionType dataModificationExceptionType = new DataModificationExceptionType(operation);
+            dataModificationExceptionType.MessageType = new MessageType { Message = errorMessage };
+
+            // This handles the Data Modification exception. This adds Core.DataModificationException annotation and also copy other instance annotations.
+            //The failed operation will be based on the protocol
+            switch (operation)
+            {
+                case DataModificationOperationKind.Update:
+                    edmChangedObject = changedObj as IEdmChangedObject;
+                    break;
+                case DataModificationOperationKind.Insert:
+                    {
+                        EdmDeltaDeletedEntityObject edmDeletedObject = new EdmDeltaDeletedEntityObject(EntityType);
+                        PatchItem(edmDeletedObject, changedObj, apiHandler);
+
+                        ValidateForDeletedEntityId(keys, edmDeletedObject);
+
+                        edmDeletedObject.TransientInstanceAnnotationContainer = changedObj.TransientInstanceAnnotationContainer;
+                        edmDeletedObject.PersistentInstanceAnnotationsContainer = changedObj.PersistentInstanceAnnotationsContainer;
+
+                        edmDeletedObject.AddDataException(dataModificationExceptionType);
+                        edmChangedObject = edmDeletedObject;
+                        break;
+                    }
+                case DataModificationOperationKind.Delete:
+                    {
+                        EdmDeltaEntityObject edmEntityObject = new EdmDeltaEntityObject(EntityType);
+                        PatchItem(originalObj as EdmStructuredObject, edmEntityObject, apiHandler);
+
+                        edmEntityObject.TransientInstanceAnnotationContainer = changedObj.TransientInstanceAnnotationContainer;
+                        edmEntityObject.PersistentInstanceAnnotationsContainer = changedObj.PersistentInstanceAnnotationsContainer;
+
+                        edmEntityObject.AddDataException(dataModificationExceptionType);
+                        edmChangedObject = edmEntityObject;
+                        break;
+                    }
+            }
+
+            return edmChangedObject;
+        }
+
+        //This is for ODL to work to set id as empty, because if there are missing keys, id wouldnt be set and we need to set it as empty.
+        private static void ValidateForDeletedEntityId(IEdmStructuralProperty[] keys, EdmDeltaDeletedEntityObject edmDeletedObject)
```

Comment:
Is `ValidateForDeletedEntityId` the appropriate name for this method? The typical implementation for `Validate...` method is to return a boolean value or to throw an exception if certain expectations are not met. This method serves a different purpose other than validating...

Reply:
We will be removing this soon since the fix in ODL has been done.